[BUG] Ingester: panic 'send on closed channel' in ActiveQueriedSeriesService — Stop races concurrent UpdateSeriesBatch senders

**AI Tool Usage Notice**

Drafted with help from Claude Code. All code references, line numbers, failure excerpts, and reproduction steps were reviewed and validated against `master` before submitting.

## Describe the bug

`ActiveQueriedSeriesService.stopping()` (`pkg/ingester/active_queried_series.go:424-430`) closes `m.updateChan` while concurrent callers of `ActiveQueriedSeriesService.UpdateSeriesBatch` (`pkg/ingester/active_queried_series.go:455-476`) can still be in the middle of sending to it. The non-blocking send uses `select { case m.updateChan <- ... : default: }`, which avoids **blocking** but does NOT protect against panicking when the channel is already closed — sends on a closed channel always panic, regardless of `select`.

Relevant code (current master):

```go
// stopping waits for all worker goroutines to finish.
func (m *ActiveQueriedSeriesService) stopping(_ error) error {
    // Close the channel to signal workers to stop
    close(m.updateChan)              // pkg/ingester/active_queried_series.go:426
    // Wait for all workers to finish
    m.workers.Wait()
    return nil
}
```

```go
// UpdateSeriesBatch sends an update to the update channel for processing.
// This method is non-blocking and will drop updates if the channel is full.
func (m *ActiveQueriedSeriesService) UpdateSeriesBatch(activeQueriedSeries *ActiveQueriedSeries, hashes []uint64, now time.Time, userID string) {
    if len(hashes) == 0 {
        return
    }
    // Non-blocking send to centralized update channel
    select {
    case m.updateChan <- activeQueriedSeriesUpdate{...}:  // pkg/ingester/active_queried_series.go:462
        // queued
    default:
        // dropped
    }
}
```

When the ingester service stops while a query path is still draining (calling `UpdateSeriesBatch` from a worker), the close + send race produces:

```
panic: send on closed channel

goroutine N [running]:
github.com/cortexproject/cortex/pkg/ingester.(*ActiveQueriedSeriesService).UpdateSeriesBatch(...)
        pkg/ingester/active_queried_series.go:462
```

Crash-on-shutdown bug; severity is operational (graceful shutdown can panic) and observable if any in-flight query goroutine outlives the moment of `close()`.

## To Reproduce

The race is timing-sensitive. Easiest reproduction is a focused unit test that runs many concurrent senders and stops the service while they are still sending. Drop the following into `pkg/ingester/active_queried_series_test.go` (or any test file in the package):

```go
func TestActiveQueriedSeriesService_StopRaceWithConcurrentSenders(t *testing.T) {
    svc := NewActiveQueriedSeriesService(log.NewNopLogger(), prometheus.NewRegistry())
    require.NoError(t, services.StartAndAwaitRunning(context.Background(), svc))

    aqs := NewActiveQueriedSeries([]time.Duration{time.Minute}, time.Second, 1.0, log.NewNopLogger())

    var wg sync.WaitGroup
    stop := make(chan struct{})
    for i := 0; i < 64; i++ {
        wg.Add(1)
        go func() {
            defer wg.Done()
            for {
                select {
                case <-stop:
                    return
                default:
                }
                // Repeatedly send; this is the call site that races with close().
                svc.UpdateSeriesBatch(aqs, []uint64{1, 2, 3, 4, 5}, time.Now(), \"user\")
            }
        }()
    }

    // Give senders a moment to fill the channel; the race window opens once
    // stopping() reaches close(m.updateChan).
    time.Sleep(10 * time.Millisecond)

    // This invokes stopping(), which closes updateChan while goroutines above
    // are still inside the select-send. Sends to a closed channel panic
    // regardless of the default branch.
    require.NoError(t, services.StopAndAwaitTerminated(context.Background(), svc))

    close(stop)
    wg.Wait()
}
```

Run it (under `-race` it tends to panic faster, but it can panic without `-race` too):

```
go test -race -tags \"netgo slicelabels\" -run TestActiveQueriedSeriesService_StopRaceWithConcurrentSenders ./pkg/ingester/... -count=50 -timeout 120s
```

Within a handful of iterations you should see:

```
--- FAIL: TestActiveQueriedSeriesService_StopRaceWithConcurrentSenders (...)
panic: send on closed channel [recovered]
    ...
goroutine ... [running]:
github.com/cortexproject/cortex/pkg/ingester.(*ActiveQueriedSeriesService).UpdateSeriesBatch(...)
        pkg/ingester/active_queried_series.go:462
```

Alternative production reproduction:
1. Start an ingester with active queried series enabled.
2. Drive sustained query load that triggers `UpdateSeriesBatch` calls.
3. Send `SIGTERM` while load is in flight. With sufficient concurrency, the ingester panics on shutdown rather than terminating cleanly.

## Expected behavior

`Stop()` should never panic, even with concurrent in-flight callers of `UpdateSeriesBatch`. Either the producer must check a \"closed\" signal before sending, or the channel must be drained / closed safely (e.g., via a `sync.Once` on close + an `atomic.Bool` flag the producer checks under a read lock, or a `context` whose cancellation gates sends).

## Suggested fix direction

Two common patterns:

1. **`atomic.Bool` + producer check** — set a `closed` flag before `close(m.updateChan)`; producers check the flag first and skip the send:
   ```go
   func (m *ActiveQueriedSeriesService) UpdateSeriesBatch(...) {
       if m.closed.Load() {
           putQueriedSeriesHashesSlice(hashes)
           return
       }
       select { case m.updateChan <- ...: default: ... }
   }
   func (m *ActiveQueriedSeriesService) stopping(_ error) error {
       m.closed.Store(true)
       close(m.updateChan)
       m.workers.Wait()
       return nil
   }
   ```
   Note: this still has a small race window between the load and the send — workable if producers also `recover()`, OR if the load+send is wrapped in a `sync.RWMutex.RLock()` paired with `Lock()` around `closed.Store(true); close(...)`.

2. **`context.Context` cancellation** — drop the channel close entirely; signal shutdown via a context, have workers select on `ctx.Done()` AND `m.updateChan` (already do), have producers select on `ctx.Done()` AND send-with-default. This avoids `close(channel)` and the inherent send-on-closed-channel hazard.

I'd suggest option 2 since `processUpdates` already selects on `ctx.Done()` — closing the channel is largely redundant.

## Environment

- Cortex master at `29b6167d6` (and reproducible on later commits).
- Found during a static audit of CPU / memory / goroutine leak risks across `pkg/`.

## Additional context

This is one of several lifecycle-race patterns surfaced by the audit; happy to send fixes if maintainers want.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -50,6 +50,7 @@
 * [BUGFIX] Tenant Federation: Fix regex resolver clearing known users list when user scan fails. #7534
 * [BUGFIX] Ingester: Fix inflight query counter leak when resource-based query protection rejects a request. #7539
 * [BUGFIX] Ingester: Release the TSDB appender on every early-return path in `Push` (e.g. out-of-order label set) by deferring `Rollback`. Previously such requests leaked TSDB head series references, mmap'd chunks and pending state per request, causing the `cortex_ingester_tsdb_head_active_appenders` gauge to grow unbounded. #7528
+* [BUGFIX] Ingester: Fix `panic: send on closed channel` in `ActiveQueriedSeriesService` on shutdown by removing the redundant channel close in `stopping()` and relying on `ctx.Done()` to signal worker exit. #7533
 * [BUGFIX] Ring: Fix ring token conflict resolution only applied to updated instance and make constantly token conflict check during instance observe period.
 * [BUGFIX] Distributor: Fix a panic (`slice bounds out of range`) in the stream push path when the context deadline expires while the worker goroutine is still marshalling a `WriteRequest`. #7541
 * [BUGFIX] Query Frontend: Fix native histogram responses not being handled correctly in `minTime()` sort ordering for split_by_interval merge. #7555

--- a/pkg/ingester/active_queried_series.go
+++ b/pkg/ingester/active_queried_series.go
@@ -421,12 +421,27 @@ func (m *ActiveQueriedSeriesService) running(ctx context.Context) error {
 }
 
 // stopping waits for all worker goroutines to finish.
+// Note: we intentionally do not close m.updateChan here. Closing it would race
+// with concurrent producers calling UpdateSeriesBatch (a select+default does
+// not prevent panics on send to a closed channel). Workers exit via ctx.Done()
+// which is signaled by the BasicService lifecycle when stopping.
 func (m *ActiveQueriedSeriesService) stopping(_ error) error {
-	// Close the channel to signal workers to stop
-	close(m.updateChan)
-	// Wait for all workers to finish
+	// Wait for all workers to finish. They will exit via ctx.Done().
 	m.workers.Wait()
-	return nil
+	// Drain any remaining updates so pooled slices are returned even if the
+	// channel was non-empty at shutdown. We never close the channel, so use a
+	// non-blocking drain. Sends that arrive after this drain exits are
+	// tolerated: UpdateSeriesBatch uses a non-blocking select+default send so
+	// producers never block, and any entries left in the buffered channel are
+	// reclaimed when the service is garbage-collected.
+	for {
+		select {
+		case update := <-m.updateChan:
+			putQueriedSeriesHashesSlice(update.hashes)
+		default:
+			return nil
+		}
+	}
 }
 
 // processUpdates is a worker goroutine that processes updates from the update channel.
@@ -437,11 +452,7 @@ func (m *ActiveQueriedSeriesService) processUpdates(ctx context.Context) {
 		select {
 		case <-ctx.Done():
 			return
-		case update, ok := <-m.updateChan:
-			if !ok {
-				// Channel closed, exit
-				return
-			}
+		case update := <-m.updateChan:
 			// Process the update synchronously
 			update.activeQueriedSeries.UpdateSeriesBatch(update.hashes, update.now)
 			// Return the slice to the pool after processing

--- a/pkg/ingester/active_queried_series_test.go
+++ b/pkg/ingester/active_queried_series_test.go
@@ -1,14 +1,19 @@
 package ingester
 
 import (
+	"context"
 	"fmt"
 	"sync"
+	"sync/atomic"
 	"testing"
 	"time"
 
 	"github.com/go-kit/log"
 	"github.com/prometheus/prometheus/model/labels"
 	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/cortexproject/cortex/pkg/util/services"
 )
 
 func TestActiveQueriedSeries_UpdateSeries(t *testing.T) {
@@ -452,4 +457,92 @@ func TestActiveQueriedSeries_EdgeCaseTimes(t *testing.T) {
 	assert.NoError(t, err)
 	// Should handle gracefully
 	assert.GreaterOrEqual(t, active, uint64(0))
+}
+
+// TestActiveQueriedSeriesService_NoSendOnClosedChannelOnShutdown is a regression
+// test for https://github.com/cortexproject/cortex/issues/7531. The service
+// previously closed updateChan in stopping() while concurrent producers were
+// still inside select+default sends, causing a "send on closed channel" panic.
+// We hammer UpdateSeriesBatch concurrently with shutdown and assert no panic.
+//
+// Each producer performs a bounded number of sends so the test always
+// terminates promptly, even when the post-stop channel-full path is exercised
+// heavily (every send takes the `default` branch and logs).
+func TestActiveQueriedSeriesService_NoSendOnClosedChannelOnShutdown(t *testing.T) {
+	const (
+		producers      = 32
+		sendsPerProd   = 5000
+		windowDuration = 1 * time.Minute
+		numWindows     = 3
+	)
+	totalDuration := time.Duration(numWindows) * windowDuration
+
+	svc := NewActiveQueriedSeriesService(log.NewNopLogger(), nil)
+	require.NoError(t, services.StartAndAwaitRunning(context.Background(), svc))
+
+	a := NewActiveQueriedSeries([]time.Duration{totalDuration}, windowDuration, 1.0, log.NewNopLogger())
+
+	// Pre-compute a hash to keep the producer loop tight and avoid expensive
+	// allocations that would dwarf the actual race window we're trying to hit.
+	hash := labels.FromStrings("a", "1").Hash()
+
+	var (
+		panicked  atomic.Bool
+		panicMsg  atomic.Value // string
+		wg        sync.WaitGroup
+		producing sync.WaitGroup
+		gate      = make(chan struct{}) // released after Stop() returns
+	)
+	wg.Add(producers)
+	producing.Add(producers)
+	for range producers {
+		go func() {
+			defer wg.Done()
+			// Capture any panic (e.g. "send on closed channel") so the test
+			// binary survives and we can fail the test cleanly.
+			defer func() {
+				if r := recover(); r != nil {
+					panicked.Store(true)
+					panicMsg.Store(fmt.Sprintf("%v", r))
+				}
+			}()
+
+			now := time.Now()
+			producing.Done()
+			// First phase: hammer while the service is still running so the
+			// channel and workers are active.
+			for range sendsPerProd {
+				hashes := getQueriedSeriesHashesSlice()
+				hashes = append(hashes, hash)
+				svc.UpdateSeriesBatch(a, hashes, now, "tenant")
+			}
+			// Wait until the test goroutine has stopped the service, then send
+			// a second burst to maximize the race window during/after
+			// shutdown. With the bug, these sends would panic.
+			<-gate
+			for range sendsPerProd {
+				hashes := getQueriedSeriesHashesSlice()
+				hashes = append(hashes, hash)
+				svc.UpdateSeriesBatch(a, hashes, now, "tenant")
+			}
+		}()
+	}
+
+	// Wait for all producers to be actively hammering before initiating
+	// shutdown, so Stop overlaps with concurrent sends.
+	producing.Wait()
+
+	// Stop the service while producers are still in their first burst. With
+	// the bug (close(updateChan) in stopping), the panic can fire here.
+	require.NoError(t, services.StopAndAwaitTerminated(context.Background(), svc))
+
+	// Release the producers' second burst — these sends target the now-stopped
+	// service and would unambiguously panic against a closed channel.
+	close(gate)
+
+	wg.Wait()
+
+	if panicked.Load() {
+		t.Fatalf("producer goroutine panicked during shutdown: %v", panicMsg.Load())
+	}
 }

--- a/pkg/ingester/active_queried_series_test.go
+++ b/pkg/ingester/active_queried_series_test.go
@@ -4,7 +4,6 @@ import (
 	"context"
 	"fmt"
 	"sync"
-	"sync/atomic"
 	"testing"
 	"time"
 
@@ -12,6 +11,7 @@ import (
 	"github.com/prometheus/prometheus/model/labels"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
+	"go.uber.org/atomic"
 
 	"github.com/cortexproject/cortex/pkg/util/services"
 )
@@ -488,7 +488,7 @@ func TestActiveQueriedSeriesService_NoSendOnClosedChannelOnShutdown(t *testing.T
 
 	var (
 		panicked  atomic.Bool
-		panicMsg  atomic.Value // string
+		panicMsg  atomic.String
 		wg        sync.WaitGroup
 		producing sync.WaitGroup
 		gate      = make(chan struct{}) // released after Stop() returns

